File: algorithms/src/PriorityQueue/PriorityQueue.cs
using System.Collections.Generic;

namespace Algorithms.PriorityQueue
{
    public class PriorityQueue<T, TPriority>
    {
        IComparer<TPriority> _priorityComparer;
        IComparer<T> _elementComparer;
        List<KeyValuePair<TPriority, T>> _entries = new List<KeyValuePair<TPriority, T>>();

        public int Count => _entries.Count - 1;

        int Root => 1;

        public PriorityQueue() : this(Comparer<TPriority>.Default, Comparer<T>.Default)
        {
        }

        public PriorityQueue(IComparer<TPriority> comparer, IComparer<T> elementComparer)
        {
            _elementComparer = elementComparer;
            _priorityComparer = comparer;
            _entries.Add(new KeyValuePair<TPriority, T>());
        }

        public void Add(T element, TPriority priority)
        {
            _entries.Add(new KeyValuePair<TPriority, T>(priority, element));
            this.HeapifyUp(_entries.Count - 1);
        }

        public KeyValuePair<TPriority, T> Peek()
        {
            return _entries[this.Root];
        }

        public KeyValuePair<TPriority, T> Pop()
        {
            var output = _entries[this.Root];
            this.Swap(this.Root, _entries.Count - 1);
            _entries.RemoveRange(_entries.Count - 1, 1);

            this.HeapifyDown(this.Root);

            return output;
        }

        public void Update(T element, TPriority oldPriority, TPriority newPriority)
        {
            for (int i = this.Root; i < _entries.Count; i++)
            {
                if (_elementComparer.Compare(_entries[i].Value, element) == 0)
                {
                    _entries[i] = new KeyValuePair<TPriority, T>(newPriority, _entries[i].Value);

                    if (_priorityComparer.Compare(newPriority, oldPriority) <= 0)
                    {
                        this.HeapifyUp(i);
                    }
                    else
                    {
                        this.HeapifyDown(i);
                    }

                    return;
                }
            }
        }

        int GetLeftChild(int index) => index * 2;

        int GetRightChild(int index) => index * 2 + 1;

        int GetParent(int index) => index / 2;

        TPriority GetPriority(int index) => _entries[index].Key;

        bool HasChildren(int index)
        {
            int left = this.GetLeftChild(index);
            int right = this.GetRightChild(index);
            int count = _entries.Count;

            return left < count || right < count;
        }

        int MaxPriorityChild(int index)
        {
            if (this.HasChildren(index))
            {
                int left = this.GetLeftChild(index);
                int right = this.GetRightChild(index);

                if (right < _entries.Count)
                {
                    if (_priorityComparer.Compare(this.GetPriority(left), this.GetPriority(right)) <= 0)
                    {
                        return left;
                    }

                    return right;
                }

                return left;
            }

            return 0;
        }

        void Swap(int a, int b)
        {
            var temp = _entries[a];
            _entries[a] = _entries[b];
            _entries[b] = temp;
        }

        void HeapifyUp(int index)
        {
            if (index == this.Root)
            {
                return;
            }

            int parent = this.GetParent(index);

            if (_priorityComparer.Compare(this.GetPriority(index), this.GetPriority(parent)) < 0)
            {
                this.Swap(index, parent);
                this.HeapifyUp(parent);
            }
        }

        void HeapifyDown(int index)
        {
            if (this.HasChildren(index))
            {
                int child = this.MaxPriorityChild(index);

                if (_priorityComparer.Compare(this.GetPriority(index), this.GetPriority(child)) > 0)
                {
                    this.Swap(child, index);
                    this.HeapifyDown(child);
                }
            }
        }
    }
}

File: python/algorithms/priority_queue.py
from typing import Generic, TypeVar, List, Callable, Tuple, Optional

T = TypeVar('T')
TPriority = TypeVar('TPriority')

class PriorityQueue(Generic[T, TPriority]):
    def __init__(
        self,
        priority_comparer: Optional[Callable[[TPriority, TPriority], int]] = None,
        element_comparer: Optional[Callable[[T, T], int]] = None
    ):
        self._priority_comparer = priority_comparer or (lambda a, b: (a > b) - (a < b))
        self._element_comparer = element_comparer or (lambda a, b: (a > b) - (a < b))
        self._entries: List[Tuple[TPriority, T]] = []  # 0-based index

    @property
    def count(self) -> int:
        return len(self._entries)

    @property
    def _root(self) -> int:
        return 0

    def add(self, element: T, priority: TPriority) -> None:
        self._entries.append((priority, element))
        self._heapify_up(len(self._entries) - 1)

    def peek(self) -> Tuple[TPriority, T]:
        return self._entries[self._root]

    def pop(self) -> Tuple[TPriority, T]:
        output = self._entries[self._root]
        self._swap(self._root, len(self._entries) - 1)
        self._entries.pop()
        self._heapify_down(self._root)
        return output

    def update(self, element: T, old_priority: TPriority, new_priority: TPriority) -> None:
        for i in range(self._root, len(self._entries)):
            if self._element_comparer(self._entries[i][1], element) == 0:
                self._entries[i] = (new_priority, self._entries[i][1])
                if self._priority_comparer(new_priority, old_priority) <= 0:
                    self._heapify_up(i)
                else:
                    self._heapify_down(i)
                return

    def _get_left_child(self, index: int) -> int:
        return 2 * index + 1

    def _get_right_child(self, index: int) -> int:
        return 2 * index + 2

    def _get_parent(self, index: int) -> int:
        return (index - 1) // 2 if index > 0 else 0

    def _get_priority(self, index: int) -> TPriority:
        return self._entries[index][0]

    def _has_children(self, index: int) -> bool:
        left = self._get_left_child(index)
        return left < len(self._entries)

    def _max_priority_child(self, index: int) -> int:
        left = self._get_left_child(index)
        right = self._get_right_child(index)
        if right < len(self._entries):
            if self._priority_comparer(self._get_priority(left), self._get_priority(right)) <= 0:
                return left
            return right
        elif left < len(self._entries):
            return left
        return -1

    def _swap(self, a: int, b: int) -> None:
        self._entries[a], self._entries[b] = self._entries[b], self._entries[a]

    def _heapify_up(self, index: int) -> None:
        if index == self._root:
            return
        parent = self._get_parent(index)
        if self._priority_comparer(self._get_priority(index), self._get_priority(parent)) < 0:
            self._swap(index, parent)
            self._heapify_up(parent)

    def _heapify_down(self, index: int) -> None:
        child = self._max_priority_child(index)
        if child != -1 and self._priority_comparer(self._get_priority(index), self._get_priority(child)) > 0:
            self._swap(child, index)
            self._heapify_down(child)
